akka-discovery/src/main/scala/akka/discovery/Discovery.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2017-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.discovery
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.function.{ Function ⇒ JFunction }
+
+import akka.actor._
+import akka.annotation.InternalApi
+
+import scala.util.{ Failure, Success }
+
+final class Discovery(implicit system: ExtendedActorSystem) extends Extension {
+
+  Discovery.checkClassPathForOldDiscovery(system)
+
+  private val implementations = new ConcurrentHashMap[String, ServiceDiscovery]
+  private val factory = new JFunction[String, ServiceDiscovery] {
+    override def apply(method: String): ServiceDiscovery = createServiceDiscovery(method)
+  }
+
+  private lazy val _defaultImplMethod =
+    system.settings.config.getString("akka.discovery.method") match {
+      case "<method>" ⇒
+        throw new IllegalArgumentException(
+          "No default service discovery implementation configured in " +
+            "`akka.discovery.method`. Make sure to configure this setting to your preferred implementation such as " +
+            "'akka-dns' in your application.conf (from the akka-discovery module).")
+      case method ⇒ method
+    }
+
+  private lazy val defaultImpl = loadServiceDiscovery(_defaultImplMethod)
+
+  /**
+   * Default [[ServiceDiscovery]] as configured in `akka.discovery.method`.
+   */
+  @throws[IllegalArgumentException]
+  def discovery: ServiceDiscovery = defaultImpl
+
+  /**
+   * Create a [[ServiceDiscovery]] from configuration property.
+   * The given `method` parameter is used to find configuration property
+   * "akka.discovery.[method].class" or "[method].class". `method` can also
+   * be a fully class name.
+   *
+   * The `ServiceDiscovery` instance for a given `method` will be created
+   * once and subsequent requests for the same `method` will return the same instance.
+   */
+  def loadServiceDiscovery(method: String): ServiceDiscovery = {
+    implementations.computeIfAbsent(method, factory)
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private def createServiceDiscovery(method: String): ServiceDiscovery = {
+    val config = system.settings.config
+    val dynamic = system.dynamicAccess
+
+    def classNameFromConfig(path: String): String =
+      if (config.hasPath(path)) config.getString(path)
+      else "<nope>"
```

Comment:
Why do we do that, 'nope'?

Reply:
Ask he who should not be named! I am guessing it was to fail loading but i'll update this to explicitly fail